### PR TITLE
Make the datasource optional for persistence units with DATABASE multi-tenancy

### LINKED ISSUE #18130: Hibernate ORM with Multitenancy and TenantConnectionResolver default datasource
## Describe the bug

When using Hibernate ORM with Multitenancy and a CustomTenantConnectionResolver like described here, I noticed that Quarkus does not start as long as I do not specify a jdbc url for the default datasource in the application.properties. Why isn't org.hibernate.engine.jdbc.connections.spi.MultiTenantConnectionProvider#getAnyConnection used for this? Or do you have to configure something else for this?

### Expected behavior
Quarkus should start and use the default datasource specified via getAnyConnection.

### Actual behavior
Quarkus does not start or you need to define the datasource in the application.properties

## To Reproduce

Link to a small reproducer (preferably a Maven project if the issue is not Gradle-specific).

Or attach an archive containing the reproducer to the issue.
[code-with-quarkus.zip](https://github.com/quarkusio/quarkus/files/6709020/code-with-quarkus.zip)

1. Without jdbc url defined start the the application -> application won't start
2. Add an jdbc url to application.properties and start the applicaytion -> application will start but uses the the datasource from application.properties and does not resolve it programmatically
3. 

### Configuration
```properties
# Add your application.properties here, if applicable.
quarkus.hibernate-orm.database.generation=none
quarkus.hibernate-orm.multitenant=DATABASE
quarkus.datasource.devservices=false

quarkus.datasource.db-kind=postgresql
#quarkus.datasource.username=postgres
#quarkus.datasource.password=test

#quarkus.datasource.jdbc.url=jdbc:postgresql://localhost:5432/postgres
```

### Screenshots
(If applicable, add screenshots to help explain your problem.)

### Output of `uname -a` or `ver`
Microsoft Windows [Version 10.0.19043.1052]

### Output of `java -version`
openjdk version "11.0.11" 2021-04-20
OpenJDK Runtime Environment AdoptOpenJDK-11.0.11+9 (build 11.0.11+9)
OpenJDK 64-Bit Server VM AdoptOpenJDK-11.0.11+9 (build 11.0.11+9, mixed mode)

### Quarkus version or git rev
1.13.7.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)
Maven 3.8.1

## Additional context
(Add any other context about the problem here.)

### PR DESCRIPTION
**Pull request by yrodiere**
## Problem

Fixes #18130 

According to [the docs](https://quarkus.io/guides/hibernate-orm#programmatically-resolving-tenants-connections), when using database multitenancy users are allowed to resolve connections manually if they so choose. However, as explained in #18130 this still requires having a default datasource around, otherwise Quarkus won't start... even if this datasource is not used by the application.

## Fix

This PR makes the datasource optional for persistence units with DATABASE multi-tenancy. It's still possible to assign a datasource to such a persistence unit (and it can be useful, see below), but it's no longer required.

## Next

There are still a few problems... most notably several warnings when the application starts.

### Agroal warning

```
2021-07-09 17:43:56,863 WARN  [io.qua.agr.dep.AgroalProcessor] (build-10) The Agroal dependency is present but no JDBC datasources have been defined.
```

Well, duh. I'm managing datasources manually. I would remove this warning, but I assume there is a good reason for it to be there?

### `JdbcEnvironmentInitiator` warning

This one is more annoying. As there is no datasource, we cannot initialize the `JdbcEnvironment` by connecting to the database. As a result we run the risk of using incorrect settings when we send SQL to the JDBC driver.

One could argue that it's a lost cause since we're connecting to multiple databases which may very well require different configuration.

But let's assume all databases require the same configurations... I can see two solutions:

1. We somehow pick one tenant to perform the auto-detection.
2. As we've discussed many times before, we try to perform the auto-detection statically, without any connection, based on the dialect or DB kind. That could be useful for other reasons (offline startup).

Either solution could work, but I think they are better suited for a separate PR.

```
2021-07-09 17:43:58,197 WARN  [org.hib.eng.jdb.env.int.JdbcEnvironmentInitiator] (JPA Startup Thread: <default>) HHH000342: Could not obtain connection to query metadata: javax.enterprise.context.ContextNotActiveException
        at io.quarkus.arc.impl.ClientProxies.getDelegate(ClientProxies.java:40)
        at io.quarkus.it.hibernate.multitenancy.fruit.CustomTenantResolver_ClientProxy.arc$delegate(CustomTenantResolver_ClientProxy.zig:59)
        at io.quarkus.it.hibernate.multitenancy.fruit.CustomTenantResolver_ClientProxy.getDefaultTenantId(CustomTenantResolver_ClientProxy.zig:176)
        at io.quarkus.hibernate.orm.runtime.tenant.HibernateMultiTenantConnectionProvider.getAnyConnectionProvider(HibernateMultiTenantConnectionProvider.java:37)
        at org.hibernate.engine.jdbc.connections.spi.AbstractMultiTenantConnectionProvider.getAnyConnection(AbstractMultiTenantConnectionProvider.java:26)
        at org.hibernate.engine.jdbc.env.internal.JdbcEnvironmentInitiator$MultiTenantConnectionProviderJdbcConnectionAccess.obtainConnection(JdbcEnvironmentInitiator.java:208)
        at org.hibernate.engine.jdbc.env.internal.JdbcEnvironmentInitiator.initiateService(JdbcEnvironmentInitiator.java:68)
        at org.hibernate.engine.jdbc.env.internal.JdbcEnvironmentInitiator.initiateService(JdbcEnvironmentInitiator.java:35)
        at org.hibernate.boot.registry.internal.StandardServiceRegistryImpl.initiateService(StandardServiceRegistryImpl.java:101)
        at org.hibernate.service.internal.AbstractServiceRegistryImpl.createService(AbstractServiceRegistryImpl.java:263)
        at org.hibernate.service.internal.AbstractServiceRegistryImpl.initializeService(AbstractServiceRegistryImpl.java:237)
        at org.hibernate.service.internal.AbstractServiceRegistryImpl.getService(AbstractServiceRegistryImpl.java:214)
        at org.hibernate.engine.jdbc.internal.JdbcServicesImpl.configure(JdbcServicesImpl.java:51)
        at org.hibernate.boot.registry.internal.StandardServiceRegistryImpl.configureService(StandardServiceRegistryImpl.java:107)
        at org.hibernate.service.internal.AbstractServiceRegistryImpl.initializeService(AbstractServiceRegistryImpl.java:246)
        at org.hibernate.service.internal.AbstractServiceRegistryImpl.getService(AbstractServiceRegistryImpl.java:214)
        at org.hibernate.boot.internal.SessionFactoryOptionsBuilder.<init>(SessionFactoryOptionsBuilder.java:263)
        at io.quarkus.hibernate.orm.runtime.recording.PrevalidatedQuarkusMetadata.buildSessionFactoryOptionsBuilder(PrevalidatedQuarkusMetadata.java:69)
        at io.quarkus.hibernate.orm.runtime.boot.FastBootEntityManagerFactoryBuilder.build(FastBootEntityManagerFactoryBuilder.java:69)
        at io.quarkus.hibernate.orm.runtime.FastBootHibernatePersistenceProvider.createEntityManagerFactory(FastBootHibernatePersistenceProvider.java:67)
        at javax.persistence.Persistence.createEntityManagerFactory(Persistence.java:80)
        at javax.persistence.Persistence.createEntityManagerFactory(Persistence.java:55)
        at io.quarkus.hibernate.orm.runtime.JPAConfig$LazyPersistenceUnit.get(JPAConfig.java:149)
        at io.quarkus.hibernate.orm.runtime.JPAConfig$1.run(JPAConfig.java:58)
        at java.base/java.lang.Thread.run(Thread.java:834)
```

### CODE REVIEW

File: integration-tests/hibernate-orm-tenancy/connection-resolver/src/main/java/io/quarkus/it/hibernate/multitenancy/AgroalRuntimeDataSourceCreationFeature.java [hunk 1]
```diff
@@ -0,0 +1,13 @@
+package io.quarkus.it.hibernate.multitenancy;
+
+import org.graalvm.nativeimage.hosted.Feature;
+import org.graalvm.nativeimage.hosted.RuntimeReflection;
+
+import io.agroal.pool.ConnectionHandler;
+
+public class AgroalRuntimeDataSourceCreationFeature implements Feature {
```
Comment (gsmet) [inline integration-tests/hibernate-orm-tenancy/connection-resolver/src/main/java/io/quarkus/it/hibernate/multitenancy/AgroalRuntimeDataSourceCreationFeature.java:8]: Is it normal that this is only done in tests? Isn't it a problem for anyone doing something similar?

File: .github/native-tests.json [hunk 2]
```diff
@@ -9,7 +9,7 @@
         {
             "category": "Data1",
             "timeout": 65,
-            "test-modules": "jpa-h2, jpa-mariadb, jpa-mssql, jpa-derby, jpa-without-entity, hibernate-orm-tenancy/datasource",
+            "test-modules": "jpa-h2, jpa-mariadb, jpa-mssql, jpa-derby, jpa-without-entity, hibernate-orm-tenancy/datasource, hibernate-orm-tenancy/connection-resolver",
```
Comment (gsmet) [inline .github/native-tests.json:12]: Not sure I would have kept this commit separated from the one where you added the test, a bit too much to my taste :).

**Review by gsmet — COMMENTED** [review]
Nice cleanup. I added a small comment and a question.

File: integration-tests/hibernate-orm-tenancy/connection-resolver/src/main/java/io/quarkus/it/hibernate/multitenancy/AgroalRuntimeDataSourceCreationFeature.java (hunk 1 above)
Comment (yrodiere) [reply integration-tests/hibernate-orm-tenancy/connection-resolver/src/main/java/io/quarkus/it/hibernate/multitenancy/AgroalRuntimeDataSourceCreationFeature.java:8]: People certainly will have the same problem if they do something similar. But we don't really recommend creating Agroal datasources at runtime, and we don't test it much either. People probably can create connections through some other ways, too.

I can move this to the Agroal extension, if you want? (using build items, of course)

File: .github/native-tests.json (hunk 2 above)
Comment (yrodiere) [reply .github/native-tests.json:12]: To each their own :) I'll squash if I need to change this PR again.

**Review by gsmet — APPROVED** [review]

File: extensions/hibernate-reactive/deployment/src/main/java/io/quarkus/hibernate/reactive/deployment/HibernateReactiveProcessor.java [hunk 3]
```diff
@@ -297,7 +302,8 @@ private static ParsedPersistenceXmlDescriptor generateReactivePersistenceUnit(
             } else if (persistenceUnitConfig.sqlLoadScript.isPresent()) {
                 //raise exception if explicit file is not present (i.e. not the default)
                 throw new ConfigurationError(
-                        "Unable to find file referenced in '" + HibernateOrmProcessor.HIBERNATE_ORM_CONFIG_PREFIX
+                        "Unable to find file referenced in '"
+                                + HibernateOrmConfig.puPropertyKey(persistenceUnitConfigName, "sql-load-script") + "="
```
Comment (Eng-Fouad) [inline extensions/hibernate-reactive/deployment/src/main/java/io/quarkus/hibernate/reactive/deployment/HibernateReactiveProcessor.java:306]: The produced error message has duplicate typo:

> Unable to find file referenced in 'quarkus.hibernate-orm.sql-load-script=sql-load-script=import.sql'. Remove property or add file to your path.

cc @yrodiere @gsmet

Comment (yrodiere) [reply extensions/hibernate-reactive/deployment/src/main/java/io/quarkus/hibernate/reactive/deployment/HibernateReactiveProcessor.java:306]: Thanks, here you go: https://github.com/quarkusio/quarkus/pull/19203

**Pull request merged**